subsys/bluetooth/controller/ll_sw/isoal.c

```diff
@@ -2163,6 +2241,208 @@ static isoal_status_t isoal_update_seg_header_cmplt_length(struct isoal_source *
+			} else {
+				/* Project a time stamp based on the last time
+				 * stamp and the number of skipped SDUs
+				 */
+				time_stamp_selected = isoal_get_wrapped_time_us(
+							session->tx_time_stamp,
+							((sdus_skipped + 1) * session->sdu_interval)
+							- session->tx_time_offset);
+				ISOAL_LOG_DBGV("[%p] Projecting Time Stamp (%lu) from skipped SDUs",
+						source_ctx, time_stamp_selected);
+			}
+
+		} else {
+			/* First SDU, align with target event */
+			if (actual_event < tx_sdu->target_event) {
+				actual_event = tx_sdu->target_event;
+				actual_grp_ref_point = tx_sdu->grp_ref_point;
+			}
+
+			ISOAL_LOG_DBGV("[%p] Use target_event", source_ctx);
+
+			if (time_stamp_is_valid) {
+				/* Time stamp is within valid range -
+				 * use provided time stamp
+				 */
+				time_stamp_selected = tx_sdu->time_stamp;
+				ISOAL_LOG_DBGV("[%p] Selecting Time Stamp (%lu) from SDU",
+						source_ctx, time_stamp_selected);
+			} else {
+				/* Time stamp is out of range -
+				 * use controller's capture time
+				 */
+				time_stamp_selected = tx_sdu->cntr_time_stamp;
+				ISOAL_LOG_DBGV("[%p] Selecting Time Stamp (%lu) from controller",
+						source_ctx, time_stamp_selected);
+			}
+		}
+
+		/* Selecting the event for transmission is done solely based on
+		 * the time stamp and the ability to calculate a valid time
+		 * offset.
+		 */
+
+		/* Check if time stamp on packet is later than the group
+		 * reference point and find next feasible event for transmission.
+		 *
+		 * BT Core V5.3 : Vol 6 Low Energy Controller : Part G IS0-AL:
+		 * 3.1 Time_Offset in framed PDUs :
+		 * The Time_Offset shall be a positive value.
+		 */
+		while (!isoal_get_time_diff(time_stamp_selected, actual_grp_ref_point, &time_diff)
+			|| time_diff == 0) {
+			/* Advance target to next event */
+			actual_event++;
+			actual_grp_ref_point = isoal_get_wrapped_time_us(actual_grp_ref_point,
+							session->iso_interval * ISO_INT_UNIT_US);
+		}
+
+		ISOAL_LOG_DBGV("[%p] Chosen PL=%llu Evt=%llu Ref=%lu",
+				source_ctx,  (actual_event * session->burst_number), actual_event,
+				actual_grp_ref_point);
+
+		/* If the event selected is the last event segmented for, then
+		 * it is possible that that some payloads have already been
+		 * released for this event. Segmentation should continue from
+		 * that payload.
+		 */
+		next_payload_number = MAX(pp->payload_number,
+					 (actual_event * session->burst_number));
+
+	ISOAL_LOG_DBGV("[%p] Final Evt=%llu (PL=%llu) Ref.=%lu Next PL=%llu",
+			source, actual_event, (actual_event * session->burst_number),
+			actual_grp_ref_point, next_payload_number);
+
+		/* Calculate the time offset */
+		time_diff_valid = isoal_get_time_diff(time_stamp_selected,
+					actual_grp_ref_point, &time_diff);
+
+		LL_ASSERT(time_diff_valid);
+		LL_ASSERT(time_diff > 0);
+		/* Time difference must be less than the maximum possible
+		 * time-offset of 24-bits.
+		 */
+		LL_ASSERT(time_diff <= 0x00FFFFFF);
```

Comment:
Wouldn't `LL_ASSERT(time_diff_valid);` be enough?

Reply:
No the time difference is valid even if it is 0 and the specification says the time offset should be a positive value. I'm not sure if 0 should be considered a positive value, but if the time offset is 0 I don't think it will make it in time for the air anyway. I am considering importing a vendor specific profiled value in deciding margins, but that's for another PR.